engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/CompositeTableDataService.java

```diff
@@ -132,43 +132,39 @@ private TableLocationProviderImpl(@NotNull final TableDataService[] inputService
             implementationName = "Composite-" + inputProviders;
 
             // Analyze the update modes of the input providers to determine the update mode of the composite provider.
-            // 1) If any providers are REFRESHING, the overall provider is REFRESHING
-            // 2) If any providers are ADD_ONLY or APPEND_ONLY, the overall provider is ADD_ONLY
-            // 3) If all providers are STATIC, the overall provider is STATIC
+            // The resultant mode is the most permissive mode of the input providers, with the exception that we will
+            // never return APPEND_ONLY.
+            boolean anyRemoves = false;
             boolean anyAdditions = false;
-            TableUpdateMode tmpMode = TableUpdateMode.STATIC;
+            boolean anyAppends = false;
             for (final TableLocationProvider provider : inputProviders) {
                 if (provider.getUpdateMode().removeAllowed()) {
-                    tmpMode = TableUpdateMode.ADD_REMOVE;
-                    break;
-                } else if (provider.getUpdateMode().addAllowed()) {
+                    anyRemoves = true;
+                } else if (provider.getLocationUpdateMode() == TableUpdateMode.ADD_ONLY) {
                     anyAdditions = true;
+                } else if (provider.getLocationUpdateMode() == TableUpdateMode.APPEND_ONLY) {
+                    anyAppends = true;
                 }
             }
```

Comment:
I think we should find a way to avoid duplicating this block. I would consolidate the two into one function, and add an append->add transformation at the end for the location (rather than location update) case.